src/com/xilinx/rapidwright/router/UltraScaleClockRouting.java

```diff
@@ -125,6 +126,117 @@ public static RouteNode routeToCentroid(Net clk, RouteNode clkRoutingLine, Clock
 		}
 		return null;		
 	}
+	
+	/**
+	 * Routes a clock from a routing track to a given transition point called the centroid
+	 * @param clk The clock net to be routed
+	 * @param startingRouteNode The starting routing track
+	 * @param centroid The given centroid node
+	 * @return
+	 */
+	public static RouteNode routeToGivenCentroidNode(Net clk, RouteNode startingRouteNode, Node centroid) {
+		Queue<RouteNode> q = new PriorityQueue<RouteNode>(16, new Comparator<RouteNode>() {
+			public int compare(RouteNode i, RouteNode j) {return i.getCost() - j.getCost();}});
+		HashSet<RouteNode> visited = new HashSet<>();
+		
+		startingRouteNode.setParent(null);
+		q.add(startingRouteNode);
+		Tile tileTarget = centroid.getTile();
+		
+		int watchDog = 10000;
+		while(!q.isEmpty()){
+			RouteNode curr = q.poll();
+			visited.add(curr);
+
+			for(Wire w : curr.getWireConnections()){
+				// Only using clk routing network to reach centroid
+				if(!w.getIntentCode().isUltraScaleClocking()) continue;
+				
+				if(w.getWireName().equals(centroid.getWireName()) 
+						&& w.getTile().equals(centroid.getTile())){
+					// curr is not the target, build the target
+					RouteNode routeNodeTarget = new RouteNode(w.getTile(), w.getWireIndex(), curr, curr.getLevel()+1);
+					clk.getPIPs().addAll(routeNodeTarget.getPIPsBackToSource());
+					return routeNodeTarget;
+				}
+					
+				RouteNode rn = new RouteNode(w.getTile(), w.getWireIndex(), curr, curr.getLevel()+1);
+				if(visited.contains(rn)) continue;
+				// using column & row based distance is more accurate than tile x/y coordinate based distance
+				int md = Math.abs(rn.getTile().getColumn() - tileTarget.getColumn()) + Math.abs(rn.getTile().getRow() - tileTarget.getRow());
+				rn.setCost(md);
+				q.add(rn);
+			}
+			if(watchDog-- == 0) {
+				throw new RuntimeException("ERROR: Could not route from " + startingRouteNode + "\n       to the given centroid: " + centroid 
+											+ ".\n       Please check if BUFGCE is correctly placed in line with the reference.");
+			}
+		}
+		return null;
+	}
+	
+	/**
+	 * Routes a clock from a routing track to a transition point where the clock 
+	 * fans out and transitions from clock routing tracks to clock distribution.
+	 * @param clk The current clock net to contribute routing
+	 * @param startingRouteNode The intermediate start point of the clock route
+	 * @param clockRegion The center clock region or the clock region that is one row above or below the center
+	 * @param findCentroidHroute The flag to indicate the returned RouteNode should be HROUTE in the center or VROUTE going up or down
+	 */
+	public static RouteNode centroidHRouteNode = null;
+	public static RouteNode routeToCentroidHRouteOrVRouteAboveBelowCentroid(Net clk, RouteNode startingRouteNode, ClockRegion clockRegion, boolean findCentroidHroute) {
```

Comment:
This method name is a bit confusing.  Can you explain what it does exactly?  It seems like it is creating the centroid in the clock network.